internal/compiler-interface/src/main/java/xsbti/compile/JavaTool.java

```diff
@@ -21,9 +21,11 @@
    *
    * @param sources  The list of java source files to compile.
    * @param options  The set of options to pass to the java compiler (includes the classpath).
+   * @param classFileManager The manager for generated class files.
    * @param reporter The reporter for semantic error messages.
    * @param log      The logger to dump output into.
    * @return true if no errors, false otherwise.
    */
-  boolean run(File[] sources, String[] options, Reporter reporter, Logger log);
+  boolean run(File[] sources, String[] options, ExternalHooks.ClassFileManager classFileManager,
```

Comment:
Ditto on this, see my earlier reply.

A new `classFileManager` is needed per incremental compile (shared by multiple cycles)

In a non-test environment, majority of time `classFileManager` is not null, it is `deleteImmediately` since transaction is only opted-in (by turning on `TransactionalManagerType` `IncOption`, in which case `deleteImmediately.generated` call is no-op.

[1] https://github.com/sbt/zinc/blob/057f52e4f6791c03c3f39cfe1e658f8f72c575da/internal/zinc-core/src/main/scala/sbt/internal/inc/ClassfileManager.scala#L54